File: src/main/java/com/github/retro_game/retro_game/controller/GalaxyController.java
package com.github.retro_game.retro_game.controller;

import com.github.retro_game.retro_game.controller.activity.Activity;
import com.github.retro_game.retro_game.dto.BuildingKindDto;
import com.github.retro_game.retro_game.dto.CoordinatesKindDto;
import com.github.retro_game.retro_game.dto.TechnologyKindDto;
import com.github.retro_game.retro_game.dto.UnitKindDto;
import com.github.retro_game.retro_game.model.ItemUtils;
import com.github.retro_game.retro_game.service.GalaxyService;
import com.github.retro_game.retro_game.service.UserService;
import com.github.retro_game.retro_game.service.impl.Converter;
import org.hibernate.validator.constraints.Range;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;

import java.time.Instant;
import java.util.Date;

@Controller
@Validated
public class GalaxyController {
  private final GalaxyService galaxyService;
  private final UserService userService;

  public GalaxyController(GalaxyService galaxyService, UserService userService) {
    this.galaxyService = galaxyService;
    this.userService = userService;
  }

  @GetMapping("/galaxy")
  @PreAuthorize("hasPermission(#bodyId, 'ACCESS')")
  @Activity(bodies = "#bodyId")
  public String galaxy(@RequestParam(name = "body") long bodyId,
                       @RequestParam @Range(min = 1, max = 5) int galaxy,
                       @RequestParam @Range(min = 1, max = 500) int system,
                       @RequestParam(required = false) @Range(min = 1, max = 15) Integer position,
                       @RequestParam(required = false) CoordinatesKindDto kind,
                       Model model) {
    var ctx = userService.getCurrentUserContext(bodyId);
    var startCoords = Converter.convert(ctx.curBody().coordinates());
    var numMissiles = ctx.curBody().units().get(UnitKindDto.INTERPLANETARY_MISSILE);

    var impulseDriveLevel = ctx.technologies().get(TechnologyKindDto.IMPULSE_DRIVE);
    var isWithinMissilesRange = ItemUtils.isWithinMissilesRange(startCoords, galaxy, system, impulseDriveLevel);

    var phalanxLevel = ctx.curBody().buildings().get(BuildingKindDto.SENSOR_PHALANX);
    var isWithinPhalanxRange = ItemUtils.isWithinPhalanxRange(startCoords, galaxy, system, phalanxLevel);

    var slots = galaxyService.getSlots(bodyId, galaxy, system);
    var now = Date.from(Instant.ofEpochSecond(Instant.now().getEpochSecond()));

    model.addAttribute("bodyId", bodyId);
    model.addAttribute("ctx", ctx);
    model.addAttribute("galaxy", galaxy);
    model.addAttribute("system", system);
    model.addAttribute("position", position);
    model.addAttribute("kind", kind);
    model.addAttribute("time", now);
    model.addAttribute("slots", slots);
    model.addAttribute("numMissiles", numMissiles);
    model.addAttribute("isWithinMissilesRange", isWithinMissilesRange);
    model.addAttribute("isWithinPhalanxRange", isWithinPhalanxRange);

    return "galaxy";
  }
}

File: src/main/kotlin/com/github/retro_game/galaxy/GalaxyController.kt
package com.github.retro_game.galaxy

import com.github.retro_game.retro_game.controller.activity.Activity
import com.github.retro_game.retro_game.dto.CoordinatesDto
import com.github.retro_game.retro_game.dto.CoordinatesKindDto
import com.github.retro_game.retro_game.dto.CoordinatesKindDto.PLANET
import com.github.retro_game.retro_game.service.PhalanxService
import com.github.retro_game.retro_game.service.UserService
import org.hibernate.validator.constraints.Range
import org.springframework.beans.factory.annotation.Value
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestParam
import java.time.Instant
import java.util.*

@Controller
@Validated
class GalaxyController constructor (
    private val galaxyService: GalaxyService,
    private val phalanxService: PhalanxService,
    private val userService: UserService,
    private val missileService: MissileService
) {

    @GetMapping("/galaxy")
    @PreAuthorize("hasPermission(#bodyId, 'ACCESS')")
    @Activity(bodies = ["#bodyId"])
    fun galaxy(
        @RequestParam(name = "body") bodyId: Long,
        @RequestParam @Range(min = 1, max = 5) galaxy: Int,
        @RequestParam @Range(min = 1, max = 500) system: Int,
        @RequestParam(required = false) @Range(min = 1, max = 15) position: Int?,
        @RequestParam(required = false) kind: CoordinatesKindDto?,
        model: Model
    ): String {
        model.apply {
            addAttribute("bodyId", bodyId)
            addAttribute("galaxy", galaxy)
            addAttribute("system", system)
            addAttribute("position", position)
            addAttribute("kind", kind)
            addAttribute("time", Date.from(Instant.ofEpochSecond(Instant.now().epochSecond)))
            addAttribute("slots", galaxyService.getSlots(galaxy, system))
            addAttribute("systemWithinRange", phalanxService.systemWithinRange(bodyId, galaxy, system))
            addAttribute("numProbes", userService.currentUserSettings.numProbes)
            addAttribute("inMissileRange", missileService.systemInRange(bodyId, galaxy, system))
        }

        return "galaxy"
    }
}
